### PR TITLE
Add a RequestDecoderHandle type which is effectively a weak reference to a RequestDecoder

### PR DESCRIPTION
**Pull request by RyanTheOptimist**
Add a RequestDecoderHandle type which is effectively a weak reference to a RequestDecoder. 
Attempt to clarify the lifetime of RequestDecoder in mobile.

Risk Level: Low - mobile only
Testing: existing tests
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by RyanTheOptimist** [comment]
/assign @RenjieTang 
WDYT? I'm not convinced that this is the right approach, but maybe?

File: envoy/http/api_listener.h [hunk 1]
```diff
@@ -5,6 +5,22 @@
...
 namespace Http {
 
+class RequestDecoderHandle {
+public:
+  virtual ~RequestDecoderHandle() = default;
+
+  /**
+   * @return true if the underlying decoder is still valid.
+   */
+  virtual bool isValid() PURE;
```
Comment (RenjieTang) [inline envoy/http/api_listener.h:15]: I think this method is never called. Maybe we can remove it?

**Review by RenjieTang — COMMENTED** [review]
Overall LGTM. A side note on better understanding the root cause:
I think we can land this change and see how it performs.
If the crash is gone, it means somewhere in this change the UAF issue is handled gracefully. We can then add some RELEASE_ASSERT with messages to figure out which request_decoder accessor is behaving badly.

File: envoy/http/api_listener.h (hunk 1 above)
Comment (RyanTheOptimist) [reply envoy/http/api_listener.h]: Good point! Done.

**Review by RenjieTang — APPROVED** [review]

**Comment by RenjieTang** [comment]
```
test/integration/api_listener_integration_test.cc:99:47: error: no member named 'newStream' in 'Envoy::Http::ApiListener'
    auto& stream_decoder = http_api_listener->newStream(stream_encoder_);
```
This looks real

**Comment by RyanTheOptimist** [comment]
> ```
> test/integration/api_listener_integration_test.cc:99:47: error: no member named 'newStream' in 'Envoy::Http::ApiListener'
>     auto& stream_decoder = http_api_listener->newStream(stream_encoder_);
> ```
> 
> This looks real

Thanks!

**Review by RenjieTang — APPROVED** [review]

**Review by abeyad — APPROVED** [review]
Great find and fix!

**Pull request merged**